### PR TITLE
fix(ai): Fix fraction seconds bug with ProtoDuration

### PR DESCRIPTION
**Pull request by daymxn**
This PR fixes a minor bug I came across in the `ProtoDuration` decoding. It's not a emergency release worthy bug, but it can produce an incorrect nanosecond value (ie; instead of `123000000` it currently produces `123`). In practicality, this means that within `LiveServerGoingAwayNotice`, the `timeLeft` field could be off by +-1 second. 

Tests have been added for `ProtoDuration` as well, to help avoid this in the future.

### CODE REVIEW

File: FirebaseAI/Sources/Types/Internal/ProtoDuration.swift [hunk 1]
```diff
@@ -107,6 +107,40 @@ extension ProtoDuration: Decodable {
...
+///
+/// // 0.123456789
+/// XCTAssertEqual(
+///   fractionalSecondsToNanoseconds(123456789, 9),
+///   123456789
+/// )
+/// ```
+private func fractionalSecondsToNanoseconds(_ value: Int32, digits: Int) -> Int32 {
+  precondition(digits >= 0 && digits <= 9, "A nanosecond value must fit within 0..9 digits")
+  precondition(value >= 0, "A nanosecond value must be positive")
```
Comment (ncooke3) [inline FirebaseAI/Sources/Types/Internal/ProtoDuration.swift]: We should avoid halting app execution over these edge cases (https://developer.apple.com/documentation/swift/precondition(_:_:file:line:)).

File: FirebaseAI/Tests/Unit/TestUtilities/XCTUtil.swift [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+/// Asserts that a string contains another string.
+///
+/// ```swift
+/// XCTAssertContains("my name is", "name")
+/// ```
+///
+/// - Parameters:
+///   - string: The source string that should contain the other.
+///   - contains: The string that should be contained in the source string.
+func XCTAssertContains(_ string: String, _ contains: String) {
```
Comment (ncooke3) [inline FirebaseAI/Tests/Unit/TestUtilities/XCTUtil.swift:26]: optional: I don't feel strongly here, but it doesn't seem like this helper is used enough to warrant it. 

Why not do something like below, inline?
```
XCAssert(error.debugDescription.contains("Invalid proto duration string"))
```

Feel free to keep it, if you'd prefer.

Comment (daymxn) [reply FirebaseAI/Tests/Unit/TestUtilities/XCTUtil.swift:26]: The error message would be less descriptive in this case, which is why I opted for a util type that explicitly logs the full "actual" value and the "expected" value. Thoughts?

File: FirebaseAI/Sources/Types/Internal/ProtoDuration.swift (hunk 1 above)
Comment (daymxn) [reply FirebaseAI/Sources/Types/Internal/ProtoDuration.swift]: SGTM, removed

File: FirebaseAI/Sources/Types/Internal/ProtoDuration.swift [hunk 3]
```diff
@@ -107,6 +107,40 @@ extension ProtoDuration: Decodable {
...
+
+  return Int32(truncatingIfNeeded: value) &* pow10[9 - digits]
```
Comment (ncooke3) [inline FirebaseAI/Sources/Types/Internal/ProtoDuration.swift]: Hmm, why can't we do `fractionalSeconds * 1_000_000_000` to get nanoseconds?

Comment (ncooke3) [reply FirebaseAI/Sources/Types/Internal/ProtoDuration.swift]: I guess we would need to parse the fractional seconds and add a leading .

File: FirebaseAI/Tests/Unit/TestUtilities/XCTUtil.swift (hunk 2 above)
Comment (ncooke3) [reply FirebaseAI/Tests/Unit/TestUtilities/XCTUtil.swift:26]: I was wondering about that. Makes sense to keep it. Resolving.

File: FirebaseAI/Sources/Types/Internal/ProtoDuration.swift [hunk 4]
```diff
@@ -107,6 +107,39 @@ extension ProtoDuration: Decodable {
...
+/// )
+///
+/// // 0.123456789
+/// XCTAssertEqual(
+///   fractionalSecondsToNanoseconds(123456789, 9),
+///   123456789
+/// )
+/// ```
+private func fractionalSecondsToNanoseconds(_ value: Int32, digits: Int) -> Int32 {
+  return Int32(truncatingIfNeeded: value) &* pow10[9 - digits]
```
Comment (ncooke3) [inline FirebaseAI/Sources/Types/Internal/ProtoDuration.swift]: Removing the preconditions is good, but should we replace them with defaults to zero? This could crash under the conditions the precondition statements meant to address.

File: FirebaseAI/Sources/Types/Internal/ProtoDuration.swift (hunk 3 above)
Comment (daymxn) [reply FirebaseAI/Sources/Types/Internal/ProtoDuration.swift]: It's not a bad idea, but I would worry about floating point precision; which is why I opted for this instead. Technically, it'd probably be fine since it fits within a double though.

File: FirebaseAI/Sources/Types/Internal/ProtoDuration.swift (hunk 4 above)
Comment (daymxn) [reply FirebaseAI/Sources/Types/Internal/ProtoDuration.swift]: Good point. Fixed.

File: FirebaseAI/Sources/Types/Internal/ProtoDuration.swift (hunk 3 above)
Comment (daymxn) [reply FirebaseAI/Sources/Types/Internal/ProtoDuration.swift]: I've adapted this, and it looks like it's working well. SGTM:)
